src/nvrtc/safe.rs

```diff
@@ -222,6 +224,180 @@ impl CompileOptions {
+}
+
+impl TryFrom<PathBuf> for PtxCrateKind {
+    type Error = String;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        if !value.exists() {
+            return Err(format!("{value:?} does not exist"))
+        }
+        
+        // if value = path/to/files
+        if let Some(name) = value
+            .file_name()
+            .map(|name| name.to_str())
+        {
+            match name {
+                // value = path/to/project/Cargo.toml
+                Some("Cargo.toml") =>
+                    return Ok( Self::Cargo { project_dir: value.parent().unwrap().into() } ),
+                // value = path/to/project/src/lib.rs
+                Some("lib.rs") => {
+                    let src = value.parent().unwrap();
+                    if let Some(project_dir) = src.parent() {
+                        let manifest = project_dir.join("Cargo.toml");
+                        if manifest.exists() {
+                            return Ok( Self::Cargo { project_dir: project_dir.into() } )
+                        } else {
+                            return Ok( Self::Standalone { standalone: value })
+                        }
+                    } else {
+                        return Err(format!("could not find parent of {src:?}"))
+                    }
+                },
+                // value = path/to/project/unsupported_name
+                Some(name) =>
+                    return Err(format!("unsupported file name: {name}")),
+                // name.to_str() failed to parse as unicode
+                None =>
+                    return Err(format!("failed to parse {name:?} as valid unicode"))
+            }
+        }
+
+        // if value = path/to/project/ containing Cargo.toml
+        if value.join("Cargo.toml").exists() {
+            Ok( Self::Cargo { project_dir: value } )
+        } else {
+            Err(format!("{value:?}/Cargo.toml missing"))
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct PtxCrate {
+    project: PtxCrateKind,
+    kernels: Option<Vec<Ptx>>,
+}
+
+impl TryInto<Vec<Ptx>> for PtxCrate {
+    type Error = String;
+    fn try_into(self) -> Result<Vec<Ptx>, Self::Error> {
+        self.kernels.ok_or("kernels not built".into())
+    }
+}
+
+impl TryFrom<PathBuf> for PtxCrate {
+    type Error = String;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        let project = value.try_into()?;
+        Ok(PtxCrate { project, kernels: None })
+    }
+}
+
+use std::process::Command;
+impl PtxCrate {
+    
+    pub fn compile_crate_to_ptx<S: AsRef<Path>>(kernel_path: S) -> Result<Vec<Ptx>, String> {
+        let kernel_path: PathBuf = kernel_path.as_ref().into();
+        let mut rust_ptx: Self = kernel_path.try_into()?;
+        rust_ptx.build_ptx()?;
+        Ok(rust_ptx.take_kernels().unwrap())
+    }
+
+    pub fn take_kernels(self) -> Option<Vec<Ptx>> {
+        self.kernels
+    }
```

Comment:
I think this has *potential* to have weird interactions with `clean()`. If someone calls .take_kernels() and then calls .clean(), the kernels would be invalidated, but the ptx objects would still be exist. Ptx::from_file just stores the file path, and only reads it on use.

We might want to separately update Ptx::from_file to read the contents on creation, or just remove this method for now.

I like this idea though, and I also like `clean`. I was actually thinking that we could do `impl Drop for PtxCrate` and call `clean`

Reply:
My thoughts were that if someone wants to take ownership of the kernels without a PtxCrate, it's on them to clean if they want. I hesitate to do `impl Drop for PtxCrate` because I fear it would interfere with `build.rs` scripts that create and then drop a `PtxCrate`.

Reply:
+1 to making it a module level free function.